pkg/ebpf/c/tracee.bpf.c

```diff
@@ -6278,4 +6253,786 @@ SEC("tc")
+CGROUP_SKB_HANDLE(proto_udp);
+CGROUP_SKB_HANDLE(proto_udp_dns);
+CGROUP_SKB_HANDLE(proto_icmp);
+CGROUP_SKB_HANDLE(proto_icmpv6);
+// clang-format on
+
+//
+// Network event submission functions
+//
+
+// clang-format off
+#define CGROUP_SKB_SUBMIT(name)                                                \
+static __always_inline u32 cgroup_skb_submit_##name(                           \
+    struct __sk_buff *ctx,                                                     \
+    u32 event_type,                                                            \
+    net_event_context_t *neteventctx,                                          \
+    u64 flags                                                                  \
+)
+// clang-format on
+
+CGROUP_SKB_SUBMIT(base)
+{
+    neteventctx->eventctx.eventid = event_type;
+    return bpf_perf_event_output(ctx,
+                                 &events,
+                                 flags,
+                                 neteventctx,
+                                 sizeof_net_event_context_t());
+}
+
+CGROUP_SKB_SUBMIT(event)
+{
+    return cgroup_skb_submit_base(ctx, event_type, neteventctx, flags);
+}
+
+CGROUP_SKB_SUBMIT(packet)
+{
+    flags = BPF_F_CURRENT_CPU;
+    flags |= (u64) ctx->len << 32; // add skb payload to perf event
+    return cgroup_skb_submit_base(ctx, event_type, neteventctx, flags);
+}
+
+//
+// SKB eBPF programs
+//
+
+static __always_inline u32 cgroup_skb_generic(struct __sk_buff *ctx)
+{
+    //
+    // MOST IMPORTANT CODE: runs for EVERY packet of a task belonging to cgroup
+    //
+
+    // use skb timestamp as the key for cgroup/skb program context
+    u64 skbts = ctx->tstamp;
+
+    net_event_context_t *neteventctx = bpf_map_lookup_elem(&cgrpctxmap, &skbts);
+    if (!neteventctx)
+        return 1;
+
+    debug_net_event_context("NEW_NET_PACKET_BASE", neteventctx);
+
+    // submit the network event only if needed
+    if (neteventctx->should_submit & SHOULD_SUBMIT_NEW_NET_PACKET)
+        cgroup_skb_submit_event(ctx, NEW_NET_PACKET, neteventctx, BPF_F_CURRENT_CPU);
+
+    // TODO: check if capture is enabled and submit full packets to userland
+
+    // fastpath: only filtering for net_packet (without capture)
+
+    u8 rest_of_submissions = SHOULD_SUBMIT_NEW_DNS_REQUEST | SHOULD_SUBMIT_NEW_DNS_RESPONSE;
+    if (!(neteventctx->should_submit & rest_of_submissions))
+        return 1; // no need to continue (save processing)
+
+    // slowpath: all subsequent operations/events need packet processing
+
+    struct bpf_sock *sk = ctx->sk;
+    if (!sk) {
+        bpf_printk("ERROR: could not get bpf_sock");
+        return 1;
+    }
+
+    sk = bpf_sk_fullsock(sk);
+    if (!sk) {
+        bpf_printk("ERROR: could not get full bpf_sock");
+        return 1;
+    }
+
+    iphdrs iphdrs = {0};
+    protohdrs protohdrs = {0};
+
+    // process protocols and needed events
+    return cgroup_skb_handle_family(ctx, neteventctx, &iphdrs, &protohdrs);
```

Comment:
I might use tail calls if needed (protocol numbers would be the index for the function/handler to call). Not needed right now (stack size is good and logic is simple).